content/5_Plat/Sparse_Segtree.mdx

```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
-	push_lazy(node);
-	if (l == segtree[node].tl && r == segtree[node].tr) return segtree[node].sum;
-	else {
-		int mid = (segtree[node].tl + segtree[node].tr) / 2;
-		if (segtree[node].l == -1) {
-			segtree[node].l = cnt++;
-			segtree[segtree[node].l].tl = segtree[node].tl;
-			segtree[segtree[node].l].tr = mid;
+	void push_down(int cur, int l, int r) {
+		if (tree[cur].left == -1) {
+			tree[cur].left = ++timer;
+			tree.push_back(Node());
+		}
+		if (tree[cur].right == -1) {
+			tree[cur].right = ++timer;
+			tree.push_back(Node());
 		}
-		if (segtree[node].r == -1) {
-			segtree[node].r = cnt++;
-			segtree[segtree[node].r].tl = mid + 1;
-			segtree[segtree[node].r].tr = segtree[node].tr;
+		int m = (l + r) / 2;
+		apply(tree[cur].left, m - l + 1, tree[cur].lazy);
+		apply(tree[cur].right, r - m, tree[cur].lazy);
+		tree[cur].lazy = 0;
+	}
+
+	void range_set(int cur, int l, int r, int ql, int qr, int val) {
+		if (qr < l || ql > r) { return; }
+		if (ql <= l && r <= qr) {
+			apply(cur, r - l + 1, val);
+		} else {
+			push_down(cur, l, r);
+			int m = (l + r) / 2;
+			range_set(tree[cur].left, l, m, ql, qr, val);
+			range_set(tree[cur].right, m + 1, r, ql, qr, val);
+			tree[cur].freq =
+			    comb(tree[tree[cur].left].freq, tree[tree[cur].right].freq);
 		}
+	}
 
-		if (l > mid) return query(segtree[node].r, l, r);
-		else if (r <= mid) return query(segtree[node].l, l, r);
-		else return query(segtree[node].l, l, mid) + query(segtree[node].r, mid + 1, r);
+	int range_sum(int cur, int l, int r, int ql, int qr) {
+		if (qr < l || ql > r) { return 0; }
+		if (ql <= l && r <= qr) { return tree[cur].freq; }
+		push_down(cur, l, r);
+		int m = (l + r) / 2;
+		return comb(range_sum(tree[cur].left, l, m, ql, qr),
+		            range_sum(tree[cur].right, m + 1, r, ql, qr));
 	}
-}
 
-int main() {
-	iostream::sync_with_stdio(false);
-	cin.tie(0);
-	int m;
-	cin >> m;
+  public:
+	SparseSegtree(int n, int q) : n(n) {
+		tree.reserve(2 * q * __lg(n));
+		tree.push_back(Node());
+	}
+
+	void range_set(int ql, int qr, int val) { range_set(0, 0, n - 1, ql, qr, val); }
 
-	segtree[1].sum = 0;
-	segtree[1].lazy = 0;
-	segtree[1].tl = 1;
-	segtree[1].tr = 1e9;
+	int range_sum(int ql, int qr) { return range_sum(0, 0, n - 1, ql, qr); }
+};
 
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	int n;
+	cin >> n;
+	SparseSegtree st((int)(1e9 + 1), n);
 	int c = 0;
-	FOR(_, 0, m) {
-		int d, x, y;
-		cin >> d >> x >> y;
-		if (d == 1) {
-			c = query(1, x + c, y + c);
+	for (int i = 0; i < n; i++) {
+		int type, x, y;
+		cin >> type >> x >> y;
+		if (type == 1) {
+			c = st.range_sum(x + c, y + c);
 			cout << c << '\n';
-		} else update(1, x + c, y + c);
+		} else if (type == 2) {
+			st.range_set(x + c, y + c, 1);
+		}
 	}
-	return 0;
 }
 ```
 
-</CPPSection>
-
-</LanguageSection>
-
 <Optional>
 
-It's possible to reduce the memory of a sparse segment tree to $\mathcal{O}(N)$,
+It's possible to reduce the memory of a sparse segment tree to $\mathcal{O}(Q)$,
```

Comment:
lmao

Reply:
$N$ is used to refer to the range size here so the prev version was wrong